custom_components/tahoma/__init__.py

```diff
@@ -36,52 +35,59 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 
     hass.data.setdefault(DOMAIN, {})
 
-    username = entry.data[CONF_USERNAME]
-    password = entry.data[CONF_PASSWORD]
+    username = entry.data.get(CONF_USERNAME)
+    password = entry.data.get(CONF_PASSWORD)
 
     try:
-        controller = await hass.async_add_executor_job(TahomaApi, username, password)
-        await hass.async_add_executor_job(controller.get_setup)
-        devices = await hass.async_add_executor_job(controller.get_devices)
-        scenes = await hass.async_add_executor_job(controller.get_action_groups)
-
-    # TODO Add better exception handling
-    except RequestException:
-        _LOGGER.exception("Error when getting devices from the TaHoma API")
+        client = TahomaClient(username, password)
+        await client.login()
+    except TooManyRequestsException as exception:
+        _LOGGER.exception(exception)
         return False
+    except BadCredentialsException as exception:
+        _LOGGER.exception(exception)
+        return False
+    except Exception as exception:  # pylint: disable=broad-except
+        _LOGGER.exception(exception)
+        return False
+
+    devices = await client.get_devices()
+    scenes = await client.get_scenarios()
+
+    hass.data[DOMAIN][entry.entry_id] = {"controller": client, "entities": {}}
 
-    hass.data[DOMAIN][entry.entry_id] = {
-        "controller": controller,
-        "devices": [],
-        "scenes": [],
-    }
+    for platform in PLATFORMS:
+        hass.data[DOMAIN][entry.entry_id]["entities"][platform] = []
+
+    for scene in scenes:
+        hass.data[DOMAIN][entry.entry_id]["entities"]["scene"].append(scene)
```

Comment:
Useless.

Reply:
Lol, good spot, this was in the original component as well... I have never spotted this as an useless loop. 👍